xbmc/platform/darwin/osx/MediaKeys.mm

```diff
@@ -0,0 +1,161 @@
+/*
+ *  Copyright (C) 2024 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "MediaKeys.h"
+
+#include "ServiceBroker.h"
+#include "input/actions/Action.h"
+#include "input/actions/ActionIDs.h"
+#include "interfaces/AnnouncementManager.h"
+#include "messaging/ApplicationMessenger.h"
+#include "utils/log.h"
+
+#import <AppKit/AppKit.h>
+#import <IOKit/hidsystem/ev_keymap.h>
+
+namespace
+{
+static CGEventRef MediaKeyCallback(CGEventTapProxy proxy,
+                                   CGEventType type,
+                                   CGEventRef event,
+                                   void* refcon)
+{
+  auto tap = (__bridge CMediaKeyTap*)refcon;
+  NSEvent* nsEvent = [NSEvent eventWithCGEvent:event];
+  if (nsEvent.type == NSEventTypeSystemDefined && nsEvent.subtype == NX_SUBTYPE_AUX_CONTROL_BUTTONS)
+  {
+    const int keyCode = (([nsEvent data1] & 0xFFFF0000) >> 16);
+    const int keyFlags = ([nsEvent data1] & 0x0000FFFF);
+    const int keyState = (((keyFlags & 0xFF00) >> 8)) == 0xA;
+    if (keyState == 1) // if pressed
+    {
+      if ([tap HandleMediaKey:keyCode])
+        return nullptr;
+    }
+  }
+  return event;
+}
+} // namespace
+
+@implementation CMediaKeyTap
+{
+  CFMachPortRef m_portRef;
+  CFRunLoopSourceRef m_sourceRef;
+  CFRunLoopRef m_tapThreadURL;
+  NSThread* m_mediaKeyTapThread;
+}
+
+- (void)dealloc
+{
+  [self disableMediaKeyTap];
+}
+
+- (void)enableMediaKeyTap
+{
+  if (m_portRef)
+    return;
+
+  m_portRef = CGEventTapCreate(kCGSessionEventTap, kCGHeadInsertEventTap, kCGEventTapOptionDefault,
+                               CGEventMaskBit(NX_SYSDEFINED), MediaKeyCallback,
+                               (__bridge void* __nullable)(self));
+  if (!m_portRef)
+  {
+    CLog::LogF(LOGERROR, "Failed to create media key tap. Check app accessibility permissions.");
+    return;
+  }
+
+  m_sourceRef = CFMachPortCreateRunLoopSource(kCFAllocatorSystemDefault, m_portRef, 0);
+  if (!m_sourceRef)
+  {
+    CLog::LogF(LOGERROR, "Failed to create media key tap. Check app accessibility permissions.");
+    return;
+  }
+
+  m_mediaKeyTapThread = [[NSThread alloc] initWithTarget:self
```

Comment:
thread object holds strong reference to target, so you essentially have reference cycle here => self will never be destroyed (without destroying the thread first). Do I understand correctly that this class basically has lifetime equal to application's lifetime, so it's not an issue in the end?

Reply:
Yes this is intended. Thread is destroyed on this class dealloc (or anytime we disable media keys). So yes, in the worst case scenario (e.g. application focused) it has lifetime equal to its parent which is the platform osx itself.